runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -1167,6 +1230,70 @@ end
 -- }}}
 -- Diagnostic User Functions {{{
 
+--- Open a floating window with the diagnostics from {position}
+---
+--- The floating window can be customized with the following highlight groups:
+--- <pre>
+--- LspDiagnosticsFloatingError
+--- LspDiagnosticsFloatingWarning
+--- LspDiagnosticsFloatingInformation
+--- LspDiagnosticsFloatingHint
+--- </pre>
+---@param opts table Configuration table
+---     - show_header (boolean, default true): Show "Diagnostics:" header.
+---@param bufnr number The buffer number
+---@param position table The (0,0)-indexed position
+---@param client_id number|nil the client id
+---@return table {popup_bufnr, win_id}
+function M.show_position_diagnostics(opts, bufnr, position, client_id)
+  opts = opts or {}
+
+  local show_header = if_nil(opts.show_header, true)
+
+  bufnr = bufnr or 0
+  if not position then
+    local curr_position = vim.api.nvim_win_get_cursor(0)
+    curr_position[1] = curr_position[1] - 1
+    position = curr_position
+  end
+
+  local lines = {}
+  local highlights = {}
+  if show_header then
+    table.insert(lines, "Diagnostics:")
+    table.insert(highlights, {0, "Bold"})
+  end
+
+  local position_diagnostics = M.get_position_diagnostics(bufnr, position, opts, client_id)
+  if vim.tbl_isempty(position_diagnostics) then return end
+
+  for i, diagnostic in ipairs(position_diagnostics) do
+    local prefix = string.format("%d. ", i)
+    local hiname = M._get_floating_severity_highlight_name(diagnostic.severity)
+    assert(hiname, 'unknown severity: ' .. tostring(diagnostic.severity))
+
+    local message_lines = vim.split(diagnostic.message, '\n', true)
+    table.insert(lines, prefix..message_lines[1])
+    table.insert(highlights, {#prefix, hiname})
+    for j = 2, #message_lines do
+      table.insert(lines, message_lines[j])
+      table.insert(highlights, {0, hiname})
+    end
+  end
+
+  opts.focus_id = "position_diagnostics"
+  local popup_bufnr, winnr = util.open_floating_preview(lines, 'plaintext', opts)
+  for i, hi in ipairs(highlights) do
+    local prefixlen, hiname = unpack(hi)
+    -- Start highlight after the prefix
+    api.nvim_buf_add_highlight(popup_bufnr, -1, hiname, i-1, prefixlen, -1)
+  end
+
+  return popup_bufnr, winnr
+end
```

Comment:
This looks pretty much like a 1:1 copy of the `show_line_diagnostics` except that it misses improvements like done in https://github.com/neovim/neovim/commit/e43dbfdd05d09e28ee0468f1cd1808f859807606 and retrieves different diagnostics.

I think you could extract the common logic related to displaying the diagnostic into an internal `show_diagnostic` function that is parameterized with the diagnostics that should be displayed and then call that function both from `show_line_diagnostics` and from this new function.

Reply:
good suggestion, I've consolidated the shared parts of both functions into `show_diagnostics`, what do you think?

Reply:
Looks good :+1